torch/distributed/_spmd/experimental_ops.py

```diff
@@ -170,6 +171,77 @@ def _prop__fused_adam(op_schema: OpSchema):
         return OutputSharding(output_spec=(op_schema.args_schema[0],) * NT)  # type: ignore[arg-type]
 
 
+@register_prop_rule(aten.nll_loss_forward.default)  # pyre-ignore
+def _prop_nll_loss_forward(op_schema: OpSchema) -> OutputSharding:
+    self, target = op_schema.args_schema[:2]
+    assert isinstance(self, DTensorSpec)
+    assert isinstance(target, DTensorSpec)
+    if self.placements != target.placements:
+        # Self and target must match in placements, which should be shard along
+        # batch dimension in data parallell use cases. Force redistribute.
+
+        # need to create a new self instead return (target, target) as target
+        # and self might not match in shape.
+        new_self = deepcopy(self)
+        new_self.placements = deepcopy(target.placements)
+        return OutputSharding(
+            output_spec=None,
+            schema_suggestions=[
+                OpSchema(
+                    func_schema=op_schema.func_schema,
+                    args_schema=(new_self, target) + op_schema.args_schema[2:],
+                    kwargs_schema=op_schema.kwargs_schema,
+                    is_inplace=op_schema.is_inplace,
+                    is_out_variant=op_schema.is_out_variant,
+                )
+            ],
+        )
+    else:
+        from torch.fx.passes.shape_prop import TensorMetadata
+
+        return OutputSharding(
+            output_spec=(
+                # by default, nll_loss_forward conducts a reduction and returns
+                # a scalar tensor, and hence the _Partial placements.
+                DTensorSpec(
+                    mesh=self.mesh,
+                    placements=[_Partial()],
```

Comment:
if we produce partial, I thought it would trigger pending reduction in the next op that consumes this partial placement during dtensor expand?

Reply:
This is the graph after DTensor expansion. I think the reason it doesn't trigger allreduce was because `nll_loss_forward` is the last operator before backward? (same as the sparse loss that @yifuwang encountered?)

```
opcode         name                        target                                   args                                                         kwargs
-------------  --------------------------  ---------------------------------------  -----------------------------------------------------------  -----------------
--------------------------------------------                                                                                                                      
placeholder    arg0_1                      arg0_1                                   ()                                                           {}
placeholder    arg0_2                      arg0_2                                   ()                                                           {}
placeholder    arg1_1                      arg1_1                                   ()                                                           {}
placeholder    arg1_2                      arg1_2                                   ()                                                           {}
placeholder    arg2_1                      arg2_1                                   ()                                                           {}               
placeholder    arg2_2                      arg2_2                                   ()                                                           {}
placeholder    arg2_3                      arg2_3                                   ()                                                           {}
placeholder    arg2_4                      arg2_4                                   ()                                                           {}
call_function  t                           aten.t.default                           (arg1_1,)                                                    {}
call_function  addmm                       aten.addmm.default                       (arg1_2, arg2_3, t)                                          {}
call_function  _log_softmax                aten._log_softmax.default                (addmm, 1, False)                                            {}
call_function  detach                      aten.detach.default                      (_log_softmax,)                                              {}
call_function  nll_loss_forward            aten.nll_loss_forward.default            (_log_softmax, arg2_4, None, 1, -100)                        {}
call_function  getitem                     <built-in function getitem>              (nll_loss_forward, 0)                                        {}
call_function  getitem_1                   <built-in function getitem>              (nll_loss_forward, 1)                                        {}
call_function  ones_like                   aten.ones_like.default                   (getitem,)                                                   {'pin_memory': False, 'memory_format': torch.preserve_format}
call_function  nll_loss_backward           aten.nll_loss_backward.default           (ones_like, _log_softmax, arg2_4, None, 1, -100, getitem_1)  {}
call_function  detach_1                    aten.detach.default                      (detach,)                                                    {}
call_function  _log_softmax_backward_data  aten._log_softmax_backward_data.default  (nll_loss_backward, detach_1, 1, torch.float32)              {}
call_function  t_1                         aten.t.default                           (_log_softmax_backward_data,)                                {}
call_function  mm                          aten.mm.default                          (t_1, arg2_3)                                                {}
call_function  t_2                         aten.t.default                           (mm,)                                                        {}
call_function  sum_1                       aten.sum.dim_IntList                     (_log_softmax_backward_data, [0], True)                      {}
call_function  view                        aten.view.default                        (sum_1, [10])                                                {}
call_function  detach_2                    aten.detach.default                      (view,)                                                      {}
call_function  detach_3                    aten.detach.default                      (detach_2,)                                                  {}
call_function  t_3                         aten.t.default                           (t_2,)                                                       {}
call_function  detach_4                    aten.detach.default                      (t_3,)                                                       {}
call_function  detach_5                    aten.detach.default                      (detach_4,)                                                  {}
call_function  all_reduce                  aten.all_reduce.default                  (detach_5, 'SUM', 'ptd:0', [0, 1], 2)                        {}
call_function  wait_tensor                 aten.wait_tensor.default                 (all_reduce,)                                                {}
call_function  all_reduce_1                aten.all_reduce.default                  (detach_3, 'SUM', 'ptd:0', [0, 1], 2)                        {}
call_function  wait_tensor_1               aten.wait_tensor.default                 (all_reduce_1,)                                              {}
call_function  _foreach_add_1              aten._foreach_add.List                   ([arg1_1, arg1_2], [wait_tensor, wait_tensor_1])             {'alpha': -0.01}
call_function  getitem_2                   <built-in function getitem>              (_foreach_add_1, 0)                                          {}
call_function  getitem_3                   <built-in function getitem>              (_foreach_add_1, 1)                                          {}
call_function  copy_                       aten.copy_.default                       (arg1_1, getitem_2)                                          {}
call_function  copy__1                     aten.copy_.default                       (arg1_2, getitem_3)                                          {}
output         output                      output                                   ([None, copy_, copy__1],)                                    {}
```

Reply:
Oh I see, this is probably because `ones_like` rule takes in a partial but it produces a replicate iirc, so it doesn't trigger allreduce.